ORST/Assets/Scripts/Core/ModuleTasks/ModuleTask.cs

```diff
@@ -0,0 +1,117 @@
+            throw new NotImplementedException();
+        }
+
+        public void StartModuleTask() {
+            if (m_CurrentModuleSubtask != null) {
+                //We have subtasks, start subtask
+                m_CurrentModuleSubtask.StartModuleTask();
+            } else {
+                Debug.Log("Task::Subtask started...");
+            }
+        }
+
+        public virtual ModuleTaskState ExecuteModuleTask() {
+            //Task implementation here, subtask will override it to implement functionality
+            return AdvanceModuleSubtasks();
+        }
+
+        private ModuleTaskState AdvanceModuleSubtasks() {
+            switch (m_CurrentModuleSubtask.ExecuteModuleTask()) {
+                case ModuleTaskState.Successful:
+                    //Subtask was successful
+                    if (m_ModuleSubtaskQueue.Count > 0) {
+                        Debug.Log("Task::Subtask successful - Advancing...");
+                        m_CurrentModuleSubtask = m_ModuleSubtaskQueue.Dequeue();
+                        if (m_CurrentModuleSubtask != null) {
+                            m_CurrentModuleSubtask.StartModuleTask();
+                            //Not all subtasks all done - task returns running
+                            return ModuleTaskState.Running;
+                        }
+                    } else {
+                        //Subtasks all done - task returns successful
+                        Debug.Log("Task::All subtasks done.");
+                        return ModuleTaskState.Successful;
+                    }
+                    break;
+
+                case ModuleTaskState.Failure:
+                    //Subtask was failure
+                    break;
+
+                case ModuleTaskState.Running:
+                    //Subtask is running
+                    break;
+
+                default:
+                    throw new SwitchExpressionException(
+                        "Task::Hit default case in 'AdvanceSubtask'. This should not happen.");
+            }
+
+            return ModuleTaskState.Failure;
+        }
+
+        public List<ModuleTask> GetRemainingModuleTasks() {
+            //Note: Not perfect since this currently loses information by flattening the hierarchy
+            if (m_ModuleSubtasks.Count <= 0) {
+                return new List<ModuleTask> { this };
+            }
+
+            //First get the current task and its remaining subtasks
+            List<ModuleTask> moduleTaskList = new() {this, m_CurrentModuleSubtask};
+            moduleTaskList.AddRange(m_CurrentModuleSubtask.GetRemainingModuleTasks());
+            //Then fetch the ones still in the queue
+            foreach (var currentModuleTask in m_ModuleSubtaskQueue) {
+                List<ModuleTask> moduleTaskTest = currentModuleTask.GetRemainingModuleTasks();
+                moduleTaskList.AddRange(moduleTaskTest);
+            }
+            return moduleTaskList;
+        }
```

Comment:
This method seems to be doing a lot of unnecessary allocations. We could replace it with something like this:
(This also fixes the issue we talked about)
```suggestion
        //Note: Not perfect since this currently loses information by flattening the hierarchy
        public IEnumerable<ModuleTask> GetRemainingModuleTasks() {
            yield return this;

            if (m_ModuleSubtasks.Count <= 0) {
                yield break;
            }

            //First get the current task and its remaining subtasks
            foreach (ModuleTask task in m_CurrentModuleSubtask.GetRemainingModuleTasks()) {
                yield return task;
            }

            //Then fetch the ones still in the queue
            foreach (ModuleTask task in m_ModuleSubtaskQueue) {
                foreach (ModuleTask subTask in task.GetRemainingModuleTasks()) {
                    yield return subTask;
                }
            }
        }
```